File: simplex.java/src/main/java/kr/co/tmon/simplex/actions/ActionBinder.java
package kr.co.tmon.simplex.actions;

class ActionBinder<TAction extends IAction<TParam, TResult>, TParam, TResult> implements IActionBinder<TAction, TParam, TResult> {

	private TAction action;
	
	public TAction getAction() {
		return action;
	}
	
//	public ActionBinder(TAction action) {
//		this.action = action;
//	}
	
	public ActionBinder(Class<TAction> clazz) throws InstantiationException, IllegalAccessException {
		action = clazz.newInstance();
	}
}

File: Simplex.kotlin/src/main/kotlin/kr/co/tmon/simplex/actions/ActionBinder.kt
package kr.co.tmon.simplex.actions

class ActionBinder<TAction : IAction<TParam, TResult>, TParam, TResult>(factory: () -> TAction)
	: IActionBinder<TAction, TParam, TResult> {

	companion object {
		inline operator fun <reified TAction : IAction<TParam, TResult>, TParam, TResult> invoke()
				= ActionBinder<TAction, TParam, TResult> {
			TAction::class.java.newInstance()
		}
	}

	override val action: TAction = factory()
}
